File: src/queue.go
package queue

import (
	"errors"
)

// Queue представляет кольцевую очередь строк
type Queue struct {
	data     []string
	capacity int
	front    int
	rear     int
	size     int
}

// New создает новую очередь с указанной емкостью
func New(capacity int) *Queue {
	return &Queue{
		data:     make([]string, capacity),
		capacity: capacity,
		front:    0,
		rear:     -1,
		size:     0,
	}
}

// NewDefault создает новую очередь с емкостью по умолчанию (30)
func NewDefault() *Queue {
	return New(30)
}

// IsEmpty проверяет, пуста ли очередь
func (q *Queue) IsEmpty() bool {
	return q.size == 0
}

// Push добавляет элемент в конец очереди
func (q *Queue) Push(value string) error {
	if q.size == q.capacity {
		return errors.New("queue is full")
	}
	
	q.rear = (q.rear + 1) % q.capacity
	q.data[q.rear] = value
	q.size++
	
	return nil
}

// Pop удаляет и возвращает элемент из начала очереди
func (q *Queue) Pop() (string, error) {
	if q.IsEmpty() {
		return "", errors.New("queue is empty")
	}
	
	value := q.data[q.front]
	q.front = (q.front + 1) % q.capacity
	q.size--
	
	return value, nil
}

// Peek возвращает элемент из начала очереди без удаления
func (q *Queue) Peek() (string, error) {
	if q.IsEmpty() {
		return "", errors.New("queue is empty")
	}
	
	return q.data[q.front], nil
}

// Size возвращает текущее количество элементов в очереди
func (q *Queue) Size() int {
	return q.size
}

// Capacity возвращает емкость очереди
func (q *Queue) Capacity() int {
	return q.capacity
}

File: src/queue.cpp
#include "../include/queue.h"

Queue::Queue(int cap) : capacity(cap), front(0), rear(-1), size(0) {
    if(capacity <= 0) {
        capacity = 30; // Защита от некорректного размера
    }
    data = new string[capacity];
}

Queue::Queue() : capacity(30), front(0), rear(-1), size(0) {
    data = new string[capacity];
}

Queue::~Queue() {
    delete[] data;
}

bool Queue::isEmpty() {
    return size == 0;
}

bool Queue::isFull() {
    return size == capacity;
}

void Queue::push(string value) {
    if (isFull()) {
        // Автоматическое расширение очереди
        int newCapacity = capacity * 2;
        string* newData = new string[newCapacity];
        
        // Копируем элементы в новый массив
        for(int i = 0; i < size; i++) {
            newData[i] = data[(front + i) % capacity];
        }
        
        delete[] data;
        data = newData;
        front = 0;
        rear = size - 1;
        capacity = newCapacity;
    }
    
    rear = (rear + 1) % capacity;
    data[rear] = value;
    size++;
}

string Queue::pop() {
    if (isEmpty()) {
        throw underflow_error("Queue is empty"); // Правильное исключение
    }
    string value = data[front];
    front = (front + 1) % capacity;
    size--;
    return value;
}

string Queue::peek() {
    if (isEmpty()) {
        throw underflow_error("Queue is empty");
    }
    return data[front];
}

int Queue::Size() {
    return size;
}

// НОВЫЙ МЕТОД: очистка очереди
void Queue::clear() {
    front = 0;
    rear = -1;
    size = 0;
}

// НОВЫЙ МЕТОД: получение емкости
int Queue::Capacity() {
    return capacity;
}
